### ISSUE TITLE
Attention Layer Bottleneck

### ISSUE THREAD

**Issue by ValterFallenius** [openclimatefix/metnet#24]
**Found a bottleneck: the attention layer**
I have found a potential bottleneck for why bug #22 occurred. It seems like the axial attention layer is some kind of bottleneck. I ran the network for 1000 epochs to try to overfit a small subset of 4 datapoints. See run at 
[WandB](https://wandb.ai/valterfallenius/lit-wandb/reports/Overfit-with-Axial-Attention--VmlldzoxNjg5MjAz?accessToken=et2bninbf8lnf8yt4m96o9p5dfk58k4xlg0n0oc02sqd0a8vi38vqo23zj9hz2ba). The network is not able to drop the loss at all almost and does not overfit the data, it yields a very bad result, some kind of mean. See image below:

<img width="922" alt="yhat against y" src="https://user-images.githubusercontent.com/21970939/158349783-89064c07-2267-4201-8f5c-404fe26a040b.PNG">

After removing the axial attention layer the model does as expected and overfits the training data, see below after 100 epochs:

<img width="926" alt="yhat against y without attention" src="https://user-images.githubusercontent.com/21970939/158350093-425fddd4-de96-4e30-8490-dbed404c4787.PNG">

The message from the author listed in #19 does mention that our implementation of axial attention seems to be very different from theirs, he says: "Our (Google's) heads were small MLPs as far as I remember (I'm not at google anymore so do not have access to the source code)." I am not experienced enough to look into the source code of our Axial Attention Library to see how this differs from theirs. 

1. What are heads in the axial attention? What does the number of heads have to do with anything?
2. Are we doing both vertical and horizontal attention passes in our implementation?

**Comment by jacobbieker** [comment]
So, this is the actual implementation being used for axial attention: https://github.com/lucidrains/axial-attention/blob/eff2c10c2e76c735a70a6b995b571213adffbbb7/axial_attention/axial_attention.py#L153-L178 which seems like ti is doing both vertical and horizontal passes. But, I just realized that we don't actually do any position embeddings, other than the lat/lon inputs, before passing to the axial attention. So we might need to add that and see what happens?  The number of heads is the number of heads for multi-headed attention. So we can probably just set it to one and be fine I think.

**Comment by jacobbieker** [comment]
Yeah, set to 1 for the debugging to get it to overfit first. And yeah, the position embedding is saying which pixel we are in, and the location information of that pixel related to other pixels in the input. The library has a function for it, so we can probably just do this for where we use the axial attention: https://github.com/openclimatefix/metnet/pull/25

**Comment by JackKelly** [comment]
>  I just realized that we don't actually do any position embeddings, other than the lat/lon inputs, before passing to the axial attention. So we might need to add that and see what happens?

I don't know if it's relevant but it recently occured to me that MetNet version 1 is quite similar to the [Temporal Fusion Transformer](https://arxiv.org/abs/1912.09363) (also from Google Research), except MetNet has 2 spatial dimensions, whilst TFT is for timeseries without any (explicit) spatial dimensions. In particular, both TFT and MetNet use an RNN followed by multi-head attention. In the TFT paper, they claim that the RNN generates a kind of learnt position encoding. So they don't bother with a "hand-crafted" position encoding.

The TFT paper says:

> [The LSTM] also serves as a replacement for standard positional encoding, providing an appropriate inductive bias for the time ordering of the inputs.

**Comment by peterdudfield** [comment]
@all-contributors please add @jacobbieker  for code

**Comment by peterdudfield** [comment]
@all-contributors please add @JackKelly  for code
